File: src/main/java/com/leon/bugreport/gui/BugReportConfirmationGUI.java
package com.leon.bugreport.gui;

import com.leon.bugreport.BugReportDatabase;
import com.leon.bugreport.BugReportLanguage;
import com.leon.bugreport.discord.LinkDiscord;
import com.leon.bugreport.keys.guiTextures;
import com.leon.bugreport.logging.ErrorMessages;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.util.List;
import java.util.*;

import static com.leon.bugreport.API.ErrorClass.logErrorMessage;
import static com.leon.bugreport.BugReportDatabase.getStaticUUID;
import static com.leon.bugreport.BugReportLanguage.getKeyFromTranslation;
import static com.leon.bugreport.BugReportManager.*;
import static com.leon.bugreport.BugReportSettings.createCustomPlayerHead;
import static com.leon.bugreport.gui.bugreportGUI.openBugReportDetailsGUI;

public class BugReportConfirmationGUI {
	public static void openConfirmationGUI(@NotNull Player player, @NotNull Boolean isArchived) {
		player.openInventory(getConfirmationGUI(isArchived));
	}

	public static @NotNull Inventory getConfirmationGUI(boolean isArchived) {
		String guiTitle;

		if (isArchived) {
			guiTitle = ChatColor.YELLOW + Objects.requireNonNull(BugReportLanguage.getValueFromLanguageFile("buttonNames.confirmationArchive", "Archive Bug Report?"));
		} else {
			guiTitle = ChatColor.YELLOW + Objects.requireNonNull(BugReportLanguage.getValueFromLanguageFile("buttonNames.confirmationDelete", "Delete Bug Report?"));
		}

		Inventory gui = Bukkit.createInventory(null, 27, guiTitle);
		ItemStack backButton = createButton(Material.BARRIER, ChatColor.YELLOW + BugReportLanguage.getValueFromLanguageFile("buttonNames.back", "Back"));
		gui.setItem(15, backButton);

		if (isArchived) {
			ItemStack archiveButton = createCustomPlayerHead(guiTextures.archiveTexture, ChatColor.YELLOW + BugReportLanguage.getValueFromLanguageFile("buttonNames.archive", "Archive"), 16);
			gui.setItem(11, archiveButton);
		} else {
			ItemStack deleteButton = createCustomPlayerHead(guiTextures.deleteTexture, ChatColor.YELLOW + BugReportLanguage.getValueFromLanguageFile("buttonNames.delete", "Delete"), 18);
			gui.setItem(11, deleteButton);
		}

		return gui;
	}

	public void archiveReport(@NotNull Player player, @NotNull Integer reportIDGUI, @NotNull Boolean isArchivedDetails) {
		BugReportDatabase.updateBugReportArchive(reportIDGUI, 1);
		player.openInventory(isArchivedDetails ? getArchivedBugReportsGUI(localCurrentPage) : getBugReportGUI(localCurrentPage));
		player.sendMessage(returnStartingMessage(ChatColor.RED)
				+ " Bug Report #" + reportIDGUI + " has been archived.");
	}

	public boolean deleteReport(@NotNull Player player, @NotNull Integer reportIDGUI, @NotNull Boolean isArchivedDetails) {
		try {
			UUID playerId = player.getUniqueId();
			BugReportDatabase.deleteBugReport(reportIDGUI);

			List<String> reports = bugReports.getOrDefault(getStaticUUID(), new ArrayList<>(Collections.singletonList("DUMMY")));
			reports.removeIf(report -> report.contains("Report ID: " + reportIDGUI));
			bugReports.put(playerId, reports);

			player.openInventory(isArchivedDetails ? getArchivedBugReportsGUI(localCurrentPage) : getBugReportGUI(localCurrentPage));
			player.sendMessage(returnStartingMessage(ChatColor.RED)
					+ " Bug Report #" + reportIDGUI + " has been deleted.");

			return true;
		} catch (Exception e) {
			return false;
		}
	}

	public record BugReportConfirmationListener(
			Inventory gui,
			Integer reportIDGUI,
			Boolean fromArchivedGUI
	) implements Listener {
		@EventHandler(priority = EventPriority.NORMAL)
		public void onInventoryClick(@NotNull InventoryClickEvent event) {
			String displayName = event.getView().getTitle();
			if (debugMode) {
				plugin.getLogger().info("Clicked inventory: " + displayName);
			}

			String customDisplayName = getKeyFromTranslation(displayName);
			if (customDisplayName == null || customDisplayName.equals(" ")) {
				return;
			}

			boolean isArchivedDetails = customDisplayName.equals("buttonNames.confirmationArchive");
			boolean isDeletedDetails = customDisplayName.equals("buttonNames.confirmationDelete");

			if (!isArchivedDetails && !isDeletedDetails) {
				String errorMessage = ErrorMessages.getErrorMessage(43);

				plugin.getLogger().severe(errorMessage);
				logErrorMessage(errorMessage);

				return;
			}

			event.setCancelled(true);

			Player player = (Player) event.getWhoClicked();
			Inventory clickedInventory = event.getClickedInventory();
			if (clickedInventory == null) {
				return;
			}

			ItemStack clickedItem = event.getCurrentItem();
			if (clickedItem == null || clickedItem.getType() == Material.AIR) {
				return;
			}

			ItemMeta itemMeta = clickedItem.getItemMeta();
			if (itemMeta == null || !itemMeta.hasDisplayName()) {
				return;
			}

			String itemDisplayName = itemMeta.getDisplayName();
			String customItemDisplayName = getKeyFromTranslation(itemDisplayName);
			if (customItemDisplayName == null || customItemDisplayName.equals(" ")) {
				return;
			}

			if (debugMode) {
				plugin.getLogger().info("Clicked item: " + customItemDisplayName);
			}

			if (isArchivedDetails) {
				if (debugMode) {
					plugin.getLogger().info("Opening archived confirmation GUI.");
				}

				switch (customItemDisplayName) {
					case "buttonNames.archive" -> {
						playButtonClickSound(player);

						if (debugMode) {
							plugin.getLogger().info("Archiving report: " + reportIDGUI);
						}
						new BugReportConfirmationGUI().archiveReport(player, reportIDGUI, true);

						if (config.getBoolean("enableDiscordWebhook", true)) {
							if (debugMode) {
								plugin.getLogger().info("Sending archive notification to Discord...");
							}

							String bugReportDiscordWebhookID = BugReportDatabase.getBugReportDiscordWebhookMessageID(reportIDGUI);
							if (bugReportDiscordWebhookID == null) {
								String errorMessage = ErrorMessages.getErrorMessage(25);

								plugin.getLogger().warning(errorMessage);
								logErrorMessage(errorMessage);
								return;
							}

							Map<String, String> fullBugReportSplit = BugReportDatabase.getBugReportById(reportIDGUI);

							String Username = fullBugReportSplit.get("Username");
							String UUID = fullBugReportSplit.get("UUID");
							String World = fullBugReportSplit.get("World");
							String FullMessage = fullBugReportSplit.get("FullMessage");
							String CategoryID = fullBugReportSplit.get("CategoryID");
							String Location = fullBugReportSplit.get("Location");
							String Gamemode = fullBugReportSplit.get("Gamemode");
							String Status = fullBugReportSplit.get("Status");
							String ServerName = fullBugReportSplit.get("ServerName");

							LinkDiscord.modifyNotification(
									Username,
									UUID,
									World,
									Location,
									Gamemode,
									Status,
									CategoryID,
									ServerName,
									FullMessage,
									bugReportDiscordWebhookID,
									Color.ORANGE,
									"Bug Report #" + reportIDGUI + " has been archived."
							);
						}

						player.openInventory(fromArchivedGUI ? getArchivedBugReportsGUI(localCurrentPage) : getBugReportGUI(localCurrentPage));

						HandlerList.unregisterAll(this);
					}
					case "buttonNames.back" -> {
						playButtonClickSound(player);

						if (debugMode) {
							plugin.getLogger().info("Going back to bug reports.");
						}
						returnFromConfirmationGUI(player);
					}
				}
			}

			if (isDeletedDetails) {
				if (debugMode) {
					plugin.getLogger().info("Opening delete confirmation GUI.");
				}

				switch (Objects.requireNonNull(customItemDisplayName)) {
					case "buttonNames.delete" -> {
						playButtonClickSound(player);

						if (debugMode) {
							plugin.getLogger().info("Deleting report: " + reportIDGUI);
						}

						String bugReportDiscordWebhookID = null;
						if (config.getBoolean("enableDiscordWebhook", true)) {
							bugReportDiscordWebhookID = BugReportDatabase.getBugReportDiscordWebhookMessageID(reportIDGUI);
						}

						Map<String, String> fullBugReportSplit = BugReportDatabase.getBugReportById(reportIDGUI);

						String Username = fullBugReportSplit.get("Username");
						String UUID = fullBugReportSplit.get("UUID");
						String World = fullBugReportSplit.get("World");
						String FullMessage = fullBugReportSplit.get("FullMessage");
						String CategoryID = fullBugReportSplit.get("CategoryID");
						String Location = fullBugReportSplit.get("Location");
						String Gamemode = fullBugReportSplit.get("Gamemode");
						String Status = fullBugReportSplit.get("Status");
						String ServerName = fullBugReportSplit.get("ServerName");

						boolean deletionSuccessful = new BugReportConfirmationGUI().deleteReport(player, reportIDGUI, isArchivedDetails);

						if (deletionSuccessful && config.getBoolean("enableDiscordWebhook", true)) {
							if (debugMode) {
								plugin.getLogger().info("Sending deletion notification to Discord...");
							}

							if (bugReportDiscordWebhookID != null) {
								LinkDiscord.modifyNotification(
										Username,
										UUID,
										World,
										Location,
										Gamemode,
										Status,
										CategoryID,
										ServerName,
										FullMessage,
										bugReportDiscordWebhookID,
										Color.RED,
										"Bug Report #" + reportIDGUI + " has been deleted."
								);
							} else {
								String errorMessage = ErrorMessages.getErrorMessage(25);
								plugin.getLogger().warning(errorMessage);
								logErrorMessage(errorMessage);
							}
						}

						player.openInventory(
								fromArchivedGUI
										? getArchivedBugReportsGUI(localCurrentPage)
										: getBugReportGUI(localCurrentPage)
						);

						HandlerList.unregisterAll(this);
					}
					case "buttonNames.back" -> {
						playButtonClickSound(player);

						if (debugMode) {
							plugin.getLogger().info("Going back to archived reports.");
						}
						returnFromConfirmationGUI(player);
					}
				}
			}
		}

		private void returnFromConfirmationGUI(@NotNull Player player) {
			player.openInventory(getBugReportGUI(localCurrentPage));

			List<String> reports = bugReports.getOrDefault(getStaticUUID(), new ArrayList<>(Collections.singletonList("DUMMY")));
			String report = reports.stream().filter(reportString -> reportString.contains("Report ID: " + reportIDGUI)).findFirst().orElse(null);

			openBugReportDetailsGUI(player, report, reportIDGUI, false);

			HandlerList.unregisterAll(this);
		}
	}
}

File: src/main/java/com/bugreportmc/bugreport/gui/BugReportConfirmationGUI.kt
package com.bugreportmc.bugreport.gui

import com.bugreportmc.bugreport.BugReportDatabase.Companion.deleteBugReport
import com.bugreportmc.bugreport.BugReportDatabase.Companion.getStaticUUID
import com.bugreportmc.bugreport.BugReportDatabase.Companion.updateBugReportArchive
import com.bugreportmc.bugreport.BugReportLanguage
import com.bugreportmc.bugreport.BugReportManager.Companion.bugReports
import com.bugreportmc.bugreport.BugReportManager.Companion.createButton
import com.bugreportmc.bugreport.BugReportManager.Companion.debugMode
import com.bugreportmc.bugreport.BugReportManager.Companion.getArchivedBugReportsGUI
import com.bugreportmc.bugreport.BugReportManager.Companion.getBugReportGUI
import com.bugreportmc.bugreport.BugReportManager.Companion.localCurrentPage
import com.bugreportmc.bugreport.BugReportManager.Companion.playButtonClickSound
import com.bugreportmc.bugreport.BugReportManager.Companion.returnStartingMessage
import com.bugreportmc.bugreport.BugReportPlugin.Companion.plugin
import com.bugreportmc.bugreport.BugReportSettings
import com.bugreportmc.bugreport.api.ErrorClass.logErrorMessage
import com.bugreportmc.bugreport.keys.guiTextures
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import java.util.*

class BugReportConfirmationGUI {
	fun archiveReport(player: Player, reportIDGUI: Int, isArchivedDetails: Boolean) {
		updateBugReportArchive(reportIDGUI, 1)
		player.openInventory(
			if (isArchivedDetails) {
				getArchivedBugReportsGUI(
					localCurrentPage, player
				)
			} else getBugReportGUI(
				localCurrentPage, player
			)
		)
		player.sendMessage(
			returnStartingMessage(ChatColor.RED) + " Bug Report #" + reportIDGUI + " has been archived."
		)
	}

	fun deleteReport(player: Player, reportIDGUI: Int, isArchivedDetails: Boolean) {
		deleteBugReport(reportIDGUI)

		val reports = bugReports.getOrDefault(
			getStaticUUID(), ArrayList<String>(listOf("DUMMY"))
		)
		reports.removeIf { report: String -> report.contains("Report ID: $reportIDGUI") }
		bugReports = bugReports.toMutableMap().apply { put(getStaticUUID(), reports) }

		player.openInventory(
			if (isArchivedDetails) getArchivedBugReportsGUI(
				localCurrentPage, player
			) else getBugReportGUI(
				localCurrentPage, player
			)
		)
		player.sendMessage(
			returnStartingMessage(ChatColor.RED) + " Bug Report #" + reportIDGUI + " has been deleted."
		)
	}

	data class BugReportConfirmationListener(
		val gui: Inventory, val reportIDGUI: Int, val fromArchivedGUI: Boolean,
	) : Listener {
		@EventHandler(priority = EventPriority.NORMAL)
		fun onInventoryClick(event: InventoryClickEvent) {
			val displayName = ChatColor.stripColor(event.view.title)

			if (debugMode) {
				plugin.logger.info("Clicked inventory: $displayName")
			}

			val customDisplayName = BugReportLanguage.getEnglishValueFromValue(displayName)
			val isArchivedDetails = customDisplayName!!.startsWith("Archive Bug Report")
			val isDeletedDetails = customDisplayName.startsWith("Delete Bug Report")

			if (!isArchivedDetails && !isDeletedDetails) {
				plugin.logger.warning("Something went wrong with the languages folder. Please remove the file and restart the server.")
				plugin.logger.warning("If the issue persists, please contact the developer.")
				logErrorMessage("Something went wrong with the languages folder.")
				return
			}

			event.isCancelled = true

			val player = event.whoClicked as Player

			val clickedItem = event.currentItem
			if (clickedItem == null || clickedItem.type == Material.AIR) {
				return
			}

			val itemMeta = clickedItem.itemMeta
			if (itemMeta == null || !itemMeta.hasDisplayName()) {
				return
			}

			val itemDisplayName = itemMeta.displayName
			val customItemDisplayName =
				BugReportLanguage.getEnglishValueFromValue(ChatColor.stripColor(itemDisplayName))

			if (debugMode) {
				plugin.logger.info("Clicked item: $customItemDisplayName")
			}

			if (isArchivedDetails) {
				if (debugMode) {
					plugin.logger.info("Opening archived confirmation GUI.")
				}
				when (customItemDisplayName) {
					"Archive" -> {
						playButtonClickSound(player)

						if (debugMode) {
							plugin.logger.info("Archiving report: $reportIDGUI")
						}
						BugReportConfirmationGUI().archiveReport(player, reportIDGUI, true)

						player.openInventory(
							if (fromArchivedGUI) getArchivedBugReportsGUI(
								localCurrentPage, player
							) else getBugReportGUI(
								localCurrentPage, player
							)
						)

						HandlerList.unregisterAll(this)
					}

					"Back" -> {
						playButtonClickSound(player)

						if (debugMode) {
							plugin.logger.info("Going back to bug reports.")
						}
						returnFromConfirmationGUI(player, false)
					}
				}
			}

			if (isDeletedDetails) {
				if (debugMode) {
					plugin.logger.info("Opening delete confirmation GUI.")
				}
				when (customItemDisplayName) {
					"Delete" -> {
						playButtonClickSound(player)

						if (debugMode) {
							plugin.logger.info("Deleting report: $reportIDGUI")
						}
						BugReportConfirmationGUI().deleteReport(player, reportIDGUI, isArchivedDetails)

						player.openInventory(
							if (fromArchivedGUI) getArchivedBugReportsGUI(
								localCurrentPage, player
							) else getBugReportGUI(
								localCurrentPage, player
							)
						)

						HandlerList.unregisterAll(this)
					}

					"Back" -> {
						playButtonClickSound(player)

						if (debugMode) {
							plugin.logger.info("Going back to archived reports.")
						}
						returnFromConfirmationGUI(player, false)
					}
				}
			}
		}

		private fun returnFromConfirmationGUI(player: Player, fromArchivedGUI: Boolean) {
			player.openInventory(
				if (fromArchivedGUI) getArchivedBugReportsGUI(
					localCurrentPage, player
				) else getBugReportGUI(
					localCurrentPage, player
				)
			)

			val reports: List<String> = bugReports.getOrDefault(
				getStaticUUID(), ArrayList<String>(listOf("DUMMY"))
			)
			val report =
				reports.stream().filter { reportString: String? -> reportString!!.contains("Report ID: $reportIDGUI") }
					.findFirst().orElse(null)

			bugreportGUI.openBugReportDetailsGUI(player, report, reportIDGUI, fromArchivedGUI)

			HandlerList.unregisterAll(this)
		}
	}

	companion object {
		fun openConfirmationGUI(player: Player, isArchived: Boolean, bugReportID: String?) {
			player.openInventory(getConfirmationGUI(isArchived, bugReportID))
		}

		private fun getConfirmationGUI(isArchived: Boolean, bugReportID: String?): Inventory {
			val guiTitle: String

			if (isArchived) {
				guiTitle = ChatColor.YELLOW.toString() + Objects.requireNonNull(
					BugReportLanguage.getValueFromLanguageFile(
						"buttonNames.confirmationArchive", "Archive Bug Report?"
					)
				)
			} else {
				guiTitle = ChatColor.YELLOW.toString() + BugReportLanguage.getValueFromLanguageFile(
					"buttonNames.confirmationDelete", "Delete Bug Report?"
				)
			}

			val gui = Bukkit.createInventory(null, 27, guiTitle)
			val backButton: ItemStack = createButton(
				Material.BARRIER,
				ChatColor.YELLOW.toString() + BugReportLanguage.getValueFromLanguageFile("buttonNames.back", "Back")
			)
			gui.setItem(15, backButton)

			if (isArchived) {
				val archiveButton = BugReportSettings.createCustomPlayerHead(
					guiTextures.archiveTexture,
					ChatColor.YELLOW.toString() + BugReportLanguage.getValueFromLanguageFile(
						"buttonNames.archive", "Archive"
					),
					16
				)
				gui.setItem(11, archiveButton)
			} else {
				val deleteButton = BugReportSettings.createCustomPlayerHead(
					guiTextures.deleteTexture, ChatColor.YELLOW.toString() + BugReportLanguage.getValueFromLanguageFile(
						"buttonNames.delete", "Delete"
					), 18
				)
				gui.setItem(11, deleteButton)
			}

			return gui
		}
	}
}
